## Data Wrangling

### Data Collection

In [2]:
# load libraries
import pandas as pd
import numpy as np
import plotly.express as px

In [5]:
# load the dataset
df = pd.read_csv("data/bank-additional-full.csv",sep=";",header=0)
df.head()

age        job  marital    education  default housing loan    contact  \
0   56  housemaid  married     basic.4y       no      no   no  telephone   
1   57   services  married  high.school  unknown      no   no  telephone   
2   37   services  married  high.school       no     yes   no  telephone   
3   40     admin.  married     basic.6y       no      no   no  telephone   
4   56   services  married  high.school       no      no  yes  telephone   

  month day_of_week  ...  campaign  pdays  previous     poutcome emp.var.rate  \
0   may         mon  ...         1    999         0  nonexistent          1.1   
1   may         mon  ...         1    999         0  nonexistent          1.1   
2   may         mon  ...         1    999         0  nonexistent          1.1   
3   may         mon  ...         1    999         0  nonexistent          1.1   
4   may         mon  ...         1    999         0  nonexistent          1.1   

   cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
0          93.994          -36.4      4.857       5191.0  no  
1          93.994          -36.4      4.857       5191.0  no  
2          93.994          -36.4      4.857       5191.0  no  
3          93.994          -36.4      4.857       5191.0  no  
4          93.994          -36.4      4.857       5191.0  no  

[5 rows x 21 columns]

### Understand the dataset

In [6]:
# check the general information about the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             41188 non-null  int64  
 1   job             41188 non-null  object 
 2   marital         41188 non-null  object 
 3   education       41188 non-null  object 
 4   default         41188 non-null  object 
 5   housing         41188 non-null  object 
 6   loan            41188 non-null  object 
 7   contact         41188 non-null  object 
 8   month           41188 non-null  object 
 9   day_of_week     41188 non-null  object 
 10  duration        41188 non-null  int64  
 11  campaign        41188 non-null  int64  
 12  pdays           41188 non-null  int64  
 13  previous        41188 non-null  int64  
 14  poutcome        41188 non-null  object 
 15  emp.var.rate    41188 non-null  float64
 16  cons.price.idx  41188 non-null  float64
 17  cons.conf.idx   41188 non-null 

There are 21 columns. Here's the snapshot of each column's description from the [data source](https://archive.ics.uci.edu/ml/datasets/bank+marketing).

Bank Client data
- age (numeric)
- job : type of job (categorical)
- marital : marital status (categorical)
- education (categorical)
- default: has credit in default? (categorical)
- housing: has housing loan? (categorical)
- loan: has personal loan? (categorical)

Related with the last contact of the current campaign:
- contact: contact communication type (categorical)
- month: last contact month of year (categorical)
- day_of_week: last contact day of the week (categorical)
- duration: last contact duration, in seconds (numeric).

Other attributes:
- campaign: number of contacts performed during this campaign and for this client (numeric, includes last contact)
- pdays: number of days that passed by after the client was last contacted from a previous campaign (numeric; 999 means client was not previously contacted)
- previous: number of contacts performed before this campaign and for this client (numeric)
- poutcome: outcome of the previous marketing campaign (categorical)

Social and economic context attributes
- emp.var.rate: employment variation rate - quarterly indicator (numeric)
- cons.price.idx: consumer price index - monthly indicator (numeric)
- cons.conf.idx: consumer confidence index - monthly indicator (numeric)
- euribor3m: euribor 3 month rate - daily indicator (numeric)
- nr.employed: number of employees - quarterly indicator (numeric)


Output variable (desired target):
- y - has the client subscribed a term deposit? (binary: 'yes','no')

The dataset looks pretty clean with no null values. Let's fix the output variable.

In [8]:
# change y to 0 and 1 instead of yes/no
df["conversion"] = df["y"].apply(lambda x: 1 if x == "yes" else 0)
df = df.drop(columns="y")
df.head()

age        job  marital    education  default housing loan    contact  \
0   56  housemaid  married     basic.4y       no      no   no  telephone   
1   57   services  married  high.school  unknown      no   no  telephone   
2   37   services  married  high.school       no     yes   no  telephone   
3   40     admin.  married     basic.6y       no      no   no  telephone   
4   56   services  married  high.school       no      no  yes  telephone   

  month day_of_week  ...  campaign  pdays  previous     poutcome emp.var.rate  \
0   may         mon  ...         1    999         0  nonexistent          1.1   
1   may         mon  ...         1    999         0  nonexistent          1.1   
2   may         mon  ...         1    999         0  nonexistent          1.1   
3   may         mon  ...         1    999         0  nonexistent          1.1   
4   may         mon  ...         1    999         0  nonexistent          1.1   

   cons.price.idx  cons.conf.idx  euribor3m  nr.employed  conversion  
0          93.994          -36.4      4.857       5191.0           0  
1          93.994          -36.4      4.857       5191.0           0  
2          93.994          -36.4      4.857       5191.0           0  
3          93.994          -36.4      4.857       5191.0           0  
4          93.994          -36.4      4.857       5191.0           0  

[5 rows x 21 columns]

In [9]:
#  check out the quality of numeric columns
df.describe()

age      duration      campaign         pdays      previous  \
count  41188.00000  41188.000000  41188.000000  41188.000000  41188.000000   
mean      40.02406    258.285010      2.567593    962.475454      0.172963   
std       10.42125    259.279249      2.770014    186.910907      0.494901   
min       17.00000      0.000000      1.000000      0.000000      0.000000   
25%       32.00000    102.000000      1.000000    999.000000      0.000000   
50%       38.00000    180.000000      2.000000    999.000000      0.000000   
75%       47.00000    319.000000      3.000000    999.000000      0.000000   
max       98.00000   4918.000000     56.000000    999.000000      7.000000   

       emp.var.rate  cons.price.idx  cons.conf.idx     euribor3m  \
count  41188.000000    41188.000000   41188.000000  41188.000000   
mean       0.081886       93.575664     -40.502600      3.621291   
std        1.570960        0.578840       4.628198      1.734447   
min       -3.400000       92.201000     -50.800000      0.634000   
25%       -1.800000       93.075000     -42.700000      1.344000   
50%        1.100000       93.749000     -41.800000      4.857000   
75%        1.400000       93.994000     -36.400000      4.961000   
max        1.400000       94.767000     -26.900000      5.045000   

        nr.employed    conversion  
count  41188.000000  41188.000000  
mean    5167.035911      0.112654  
std       72.251528      0.316173  
min     4963.600000      0.000000  
25%     5099.100000      0.000000  
50%     5191.000000      0.000000  
75%     5228.100000      0.000000  
max     5228.100000      1.000000

### Data Cleaning to-dos so far
- Need to change column names to something more meaningful
- We need to clean up "pdays" to remove 999, since it affects the distribution of this attribute
- Why are some index and rate negative? Do we need to fix them before modeling?
- We need to check the quality of the categorical columns to see if we need to do string cleaning

We will come back to data cleaning later. For now we will focus on descriptive analysis of Conversion Rate

### Descriptive Analysis of Conversion Rate

In [15]:
# total number of conversions out of total clients contacted
print(f"Number of conversions: {df.conversion.sum()} out of {df.conversion.count()}.")

Number of conversions: 4640 out of 41188.


In [20]:
# conversion rate
print(f"Conversion rate: {round(df.conversion.sum()/df.conversion.count()*100,2)}%")

Conversion rate: 11.27%


In [23]:
# conversion by age groups
# create age groups
df["age_group"] = df["age"].apply(
    lambda x: "18-24" if x <= 24
    else "25-30" if x <= 30
    else "31-45" if x <= 45
    else "45-60" if x <= 60
    else "60-75" if x <=75 else "75+")
df["age_group"].value_counts()

31-45    21974
45-60    10921
25-30     6315
18-24     1068
60-75      665
75+        245
Name: age_group, dtype: int64

In [31]:
# calculate conversion rate for each age group
conversion_by_agegroup = round(df.groupby(by="age_group")["conversion"].sum()/df.groupby(by="age_group")["conversion"].count()*100,2).to_frame().reset_index()
print(conversion_by_agegroup)

  age_group  conversion
0     18-24       23.97
1     25-30       13.75
2     31-45        9.37
3     45-60        9.56
4     60-75       42.86
5       75+       52.65


In [33]:
# plot on bar chart
fig = px.bar(conversion_by_agegroup, x="age_group",y="conversion", title="Conversion rate by age group", text="conversion")
fig.show()

#### Comments
- Senior age groups have a very high conversion rate, but low volume.
- Age groups with highest number of customers have lowest conversion rates (age 31-45 and 45-60)

In [34]:
# check conversion rate by job
df["job"].value_counts()

admin.           10422
blue-collar       9254
technician        6743
services          3969
management        2924
retired           1720
entrepreneur      1456
self-employed     1421
housemaid         1060
unemployed        1014
student            875
unknown            330
Name: job, dtype: int64

In [37]:
# calculate conversion rate for each job
conversion_by_job = round(df.groupby(by="job")["conversion"].sum()/df.groupby(by="job")["conversion"].count()*100,2).to_frame().reset_index()

#plot on bar chat
fig = px.bar(conversion_by_job, x="job",y="conversion", title="Conversion rate by job", text="conversion")
fig.show()

#### Conversion vs. Non-conversion

In [72]:
# number of conversions and non-conversions by education
conversions_by_education = pd.pivot_table(df, values="contact",index = "education",columns="conversion",aggfunc=len)
conversions_by_education = conversions_by_education.rename(columns={1:"conversion",0:"non_conversion"}).reset_index().melt(id_vars="education",value_vars=["conversion","non_conversion"])
conversions_by_education

education      conversion  value
0              basic.4y      conversion    428
1              basic.6y      conversion    188
2              basic.9y      conversion    473
3           high.school      conversion   1031
4            illiterate      conversion      4
5   professional.course      conversion    595
6     university.degree      conversion   1670
7               unknown      conversion    251
8              basic.4y  non_conversion   3748
9              basic.6y  non_conversion   2104
10             basic.9y  non_conversion   5572
11          high.school  non_conversion   8484
12           illiterate  non_conversion     14
13  professional.course  non_conversion   4648
14    university.degree  non_conversion  10498
15              unknown  non_conversion   1480

In [74]:
fig = px.bar(conversions_by_education,x = "education",y="value",color="conversion",barmode="group",
            title="Conversions and non-conversions by education"
            )
fig.show()
